### PR TITLE
usaco dishwashing lololol

### PR DESCRIPTION
**Pull request by envyaims**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/usaco-922.mdx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main(){
+	ifstream cin("dishes.in");
+	ofstream cout("dishes.out");
+	
+	int n; cin >> n;
```
Comment (SansPapyrus683) [inline solutions/usaco-922.mdx:26]: separate lines

File: solutions/usaco-922.mdx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main(){
+	ifstream cin("dishes.in");
+	ofstream cout("dishes.out");
```
Comment (SansPapyrus683) [inline solutions/usaco-922.mdx:24]:
```suggestion
	ifstream cin("dishes.in");
```
please rename these to something else

File: solutions/usaco-922.mdx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main(){
```
Comment (SansPapyrus683) [inline solutions/usaco-922.mdx:22]:
```suggestion
int main() {
```

File: solutions/usaco-922.mdx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	while (l < r - 1) {
+		int mid = (l + r) / 2;
+		if (check(mid)) {
+			l = mid;
+		}
+		else {
+			r = mid;
+		}
+	}
+	cout << l << endl;
```
Comment (SansPapyrus683) [inline solutions/usaco-922.mdx:93]:
```suggestion
	std::ofstream("dishes.out") << l << endl;
```

File: solutions/usaco-922.mdx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+			int plate = order[i];
+			
+			// binary search for the first stack it can be placed
+			int l = -1, r = soapy_stacks.size();
+			while (l < r - 1) {
+				int mid = (l + r) / 2;
+				if (soapy_stacks[mid].back() > plate) {
+					r = mid;
+				}
+				else{
```
Comment (SansPapyrus683) [inline solutions/usaco-922.mdx:53]:
```suggestion
				} else {
```
same for other else blocks

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
honestly pretty well commented code

File: solutions/usaco-922.mdx (hunk 5 above)
Comment (envyaims) [reply solutions/usaco-922.mdx:53]: i dont get how do people do this
this is literally war crimes

File: solutions/usaco-922.mdx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	ifstream fin("dishes.in");
+	int n; 
+	fin >> n;
+	vector<int> order(n);
+	for (int i = 0; i < n; i++) {
+		fin >> order[i];
+	}
+	
+	// returns true if dishes with indexes [0, end_index) can be washed
+	auto check = [&] (int end_index) -> bool {
```
Comment (SansPapyrus683) [inline solutions/usaco-922.mdx:32]: maybe rename to `prefix_washable`

File: solutions/usaco-922.mdx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+			// not able to be placed in any existing stack 
+			if (r == soapy_stacks.size()) {
+				// create new stack
+				soapy_stacks.push_back({plate});
+			} else {
+				soapy_stacks[r].push_back(plate);
+			}
+			
+			// while are able to wash a plate, wash it
+			while (!soapy_stacks.empty() && soapy_stacks.front().back() == not_washed.front()) {
```
Comment (SansPapyrus683) [inline solutions/usaco-922.mdx:65]: over 80 chars? not sure

File: solutions/usaco-922.mdx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+source: USACO Gold 2019 February
+title: Dishwashing
+author: Chongtian Ma
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_dishes_gold_feb19.html)
+
+## Alternate Solution - Binary Search
+
+We can binary search for the longest possible prefix. Prefix of length $i-1$ is just prefix of length $i$ without a plate; if is it possible to wash all plates of prefix $i$ then it must also be possible to wash a prefix of $i-1$. Now we only need to simulate the stacking and washing. As the official analysis outlines, if plate $A$ is on top of plate $B$ on the same stack, then $A$ must be less than $B$, or else we cannot access the plate beneath it. Thus, we can greedily place each plate on stack $i$ if the top plate on stack $i$ is larger than our current plate and is the smallest out of all the other soapy stacks. This is because it is always more optimal to satisfy a stricter constraint, rather overriding a broader one if we place the plate on another stack which the plate on top is not the smallest possible.
```
Comment (SansPapyrus683) [inline solutions/usaco-922.mdx:12]: can you break this up into paragraphs

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-922.mdx (hunk 8 above)
Comment (envyaims) [reply solutions/usaco-922.mdx:12]: what 
its not even that much

File: solutions/usaco-922.mdx (hunk 7 above)
Comment (envyaims) [reply solutions/usaco-922.mdx]: dammit 84

File: solutions/usaco-922.mdx (hunk 8 above)
Comment (SansPapyrus683) [reply solutions/usaco-922.mdx:12]: it's really really hard for me to review a giant line of text

Comment (envyaims) [reply solutions/usaco-922.mdx]: ok sans just for you <3

File: solutions/usaco-922.mdx [hunk 9]
```diff
@@ -9,7 +9,9 @@ author: Chongtian Ma
...
+We can binary search for the longest possible prefix. Prefix of length $i-1$ is just prefix of length $i$ without a plate; if is it possible to wash all plates of prefix $i$ then it must also be possible to wash a prefix of $i-1$. Now we only need to simulate the stacking and washing. As the official analysis outlines, if plate $A$ is on top of plate $B$ on the same stack, then $A$ must be less than $B$, or else we cannot access the plate beneath it.
```
Comment (SansPapyrus683) [inline solutions/usaco-922.mdx:12]: > Prefix of length $i-1$ is just prefix of length $i$ without a plate;

grammar

Comment (envyaims) [reply solutions/usaco-922.mdx:12]: wait whats wron gwith it lol

Comment (SansPapyrus683) [reply solutions/usaco-922.mdx:12]: it's a fragment?

Comment (envyaims) [reply solutions/usaco-922.mdx:12]: wha can u just fix it for me lol

Comment (SansPapyrus683) [reply solutions/usaco-922.mdx:12]: no.

Comment (envyaims) [reply solutions/usaco-922.mdx:12]: ill just assume nothing is wrong then unless someone says otherwise 😄

Comment (SansPapyrus683) [reply solutions/usaco-922.mdx:12]: how is a fragment not wrong

Comment (envyaims) [reply solutions/usaco-922.mdx:12]: am i blind wheres the fragment

Comment (SansPapyrus683) [reply solutions/usaco-922.mdx:12]: https://en.wikipedia.org/wiki/Semicolon

> a semicolon is most commonly used to link (in a single sentence) two independent clauses that are closely related in thought

> Prefix of length $i-1$ is just prefix of length $i$ without a plate;

you tell me how this is an independent clause

Comment (envyaims) [reply solutions/usaco-922.mdx:12]: "Prefix" is the subject, "is" is the verb? Do you want me to add "A" in front of it or something

Comment (SansPapyrus683) [reply solutions/usaco-922.mdx:12]: oh yeah, mb mb
what i meant is "Prefix" being the subject is very very awkward

**Review by ryanchou-dev — APPROVED** [review]
lolol

File: solutions/usaco-922.mdx (hunk 9 above)
Comment (SansPapyrus683) [reply solutions/usaco-922.mdx:12]: not addressed

Comment (envyaims) [reply solutions/usaco-922.mdx:12]: fine ill add your "A"

File: content/4_Gold/Stacks.problems.json [hunk 10]
```diff
@@ -58,10 +58,9 @@
       "source": "Gold",
       "difficulty": "Normal",
       "isStarred": false,
-      "tags": ["Stack"],
+      "tags": ["Stack, Binary Search"],
```
Comment (SansPapyrus683) [inline content/4_Gold/Stacks.problems.json:61]: you did the tags wrong

Comment (envyaims) [reply content/4_Gold/Stacks.problems.json:61]: que?

Comment (SansPapyrus683) [reply content/4_Gold/Stacks.problems.json:61]: it's supposed to be 2 strings, you put them as 1

Comment (envyaims) [reply content/4_Gold/Stacks.problems.json:61]: bruh u cant tell on the website but ok

**Pull request merged**